client/pages/room.jsx

```diff
@@ -25,28 +32,82 @@ export default class Room extends React.Component {
       });
   }
 
+  handleUsernameInput(event) {
+    const target = event.target;
+    const value = target.value;
+    const name = target.name;
+    this.setState({ [name]: value });
+    this.props.handleUsernameInput(event.target.value);
+  }
+
+  handleSubmit(event) {
+    event.preventDefault();
+    const req = {
+      method: 'POST',
+      headers: {
+        'Content-Type': 'application/json'
+      },
+      body: JSON.stringify(this.state)
+    };
+    fetch('/api/user/username', req)
+      .then(res => res.json())
+      .then(result => {
+        this.setState({
+          username: result.username
+        });
+      });
+    this.setState(prevState => ({
+      modal: !prevState.modal
+    }));
```

Comment:
It's best to do "clean up" steps when you know that your asynchronous stuff is done (i.e. in your second `.then()` callback

```js
      .then(result => {
        this.setState({
          modal: false,
          username: result.username
        });
      });
```